src/toil/test/src/threadingTest.py

```diff
@@ -0,0 +1,145 @@
+
+class ThreadingTest(ToilTest):
+    """
+    Test Toil threading/synchronization tools
+    """
+    
+    @travis_test
+    def testGlobalMutexOrdering(self):
+        for it in range(10):
+            log.info('Iteration %d', it)
+        
+            scope = self._createTempDir()
+            mutex = 'mutex'
+            # Use processes (as opposed to threads) to prevent GIL from ordering things artificially
+            pool = multiprocessing.Pool()
+            try:
+                numTasks = 100
+                results = pool.map_async(
+                    func=partial(_testGlobalMutexOrderingTask, scope, mutex),
+                    iterable=list(range(numTasks)))
+                results = results.get()
+            finally:
+                pool.close()
+                pool.join()
+            
+            self.assertEqual(len(results), numTasks)
+            for item in results:
+                # Make sure all workers say they succeeded
+                self.assertEqual(item, True)
+                
+    @travis_test
+    def testLastProcessStanding(self):
+        for it in range(10):
+            log.info('Iteration %d', it)
+        
+            scope = self._createTempDir()
+            arena_name = 'thunderdome'
+            # Use processes (as opposed to threads) to prevent GIL from ordering things artificially
+            pool = multiprocessing.Pool()
+            try:
+                numTasks = 100
+                results = pool.map_async(
+                    func=partial(_testLastProcessStandingTask, scope, arena_name),
+                    iterable=list(range(numTasks)))
+                results = results.get()
+            finally:
+                pool.close()
+                pool.join()
+            
+            self.assertEqual(len(results), numTasks)
+            for item in results:
+                # Make sure all workers say they succeeded
+                self.assertEqual(item, True)
+            for filename in os.listdir(scope):
+                assert not filename.startswith('precious'), "File {} still exists".format(filename)
+
+def _testGlobalMutexOrderingTask(scope, mutex, number):
+    try:
+        # We will all fight over the potato
+        potato = os.path.join(scope, 'potato')
+        
+        with global_mutex(scope, mutex):
+            log.info('PID %d = num %d running', os.getpid(), number)
+            assert not os.path.exists(potato), "We see someone else holding the potato file"
+            
+            # Put our name there
+            out_stream = open(potato, 'w')
+            out_stream.write(str(number))
+            out_stream.close()
+            
+            # Wait
+            time.sleep(random.random() * 0.01)
+            
+            # Make sure our name is still there
+            in_stream = open(potato, 'r')
+            seen = in_stream.read().rstrip() 
+            assert seen == str(number), "We are {} but {} stole our potato!".format(number, seen)
```

Comment:
I'm not sure how python handles open file handles upon an exception exit these days, but could I suggest either moving the `in_stream.close()` above the assert statement, or using the context manager `with open(file, 'r') as f:` (possibly for the code several lines above it as well where it will mostly just save a line of code).

Reply:
That makes sense.